test/integration/consul-container/test/tproxy/tproxy_test.go

```diff
@@ -30,34 +33,74 @@ import (
 func TestTProxyService(t *testing.T) {
 	t.Parallel()
 
-	cluster, _, _ := topology.NewCluster(t, &topology.ClusterConfig{
-		NumServers:                1,
-		NumClients:                2,
-		ApplyDefaultProxySettings: true,
-		BuildOpts: &libcluster.BuildOptions{
-			Datacenter:             "dc1",
-			InjectAutoEncryption:   true,
-			InjectGossipEncryption: true,
-			// TODO(rb): fix the test to not need the service/envoy stack to use :8500
-			AllowHTTPAnyway: true,
-		},
-	})
+	cluster := createCluster(t, 2) // 2 client agent pods
 
 	clientService := createServices(t, cluster)
 	_, adminPort := clientService.GetAdminAddr()
 
 	libassert.AssertUpstreamEndpointStatus(t, adminPort, "static-server.default", "HEALTHY", 1)
 	libassert.AssertContainerState(t, clientService, "running")
-	assertHTTPRequestToVirtualAddress(t, clientService)
+	assertHTTPRequestToStaticServerVirtual(t, clientService)
 }
 
-func assertHTTPRequestToVirtualAddress(t *testing.T, clientService libservice.Service) {
-	timer := &retry.Timer{Timeout: 120 * time.Second, Wait: 500 * time.Millisecond}
+// TestTProxyPermissiveMTLS makes sure that a service in permissive mTLS mode accepts
+// non-mesh traffic on the upstream service's port.
+//
+// Steps:
+//   - Create a single server cluster
+//   - Create the static-server and static-client services in the mesh
+//   - Create an additional non-mesh container without a mesh service
+//   - In default/strict mTLS mode, make sure requests from static-client to
+//     static-server's virtual address succeed, but requests from the non-mesh
+//     service fail
+//   - In permissive mTLS mode, make sure requests from static-client to
+//     static-server's virtual addresss succeed, and requests from the non-mesh
+//     service to the static-server's normal address/port succeed.
+func TestTProxyPermissiveMTLS(t *testing.T) {
+	t.Parallel()
+
+	// Three client containers: static-client, static-server, and a non-mesh service.
+	// To simulate requests from outside the mesh, we use another client agent container
+	// without a mesh service, and run curl to simulate requests from outside the mesh.
+	cluster := createCluster(t, 3)
+
+	staticServerPod := cluster.Agents[1]
```

Comment:
Do we use `cluster.Agents[2]` at all?

Reply:
Yes. Each cluster.Agent is a sort of simulated "pod". 

* cluster.Agents[0] - The Consul server
* cluster.Agents[1] - The "static-server" service
* cluster.Agents[2] - The "static-client" service ([link](https://github.com/hashicorp/consul/blob/af75461e90286f253f8d4da901e177133f2410f9/test/integration/consul-container/test/tproxy/tproxy_test.go#L200-L210))
* cluster.Agents[3] - Another pod where the test can make requests from a non-mesh service.